packages/protocol/contracts/governance/DowntimeSlasher.sol

```diff
@@ -35,88 +42,212 @@ contract DowntimeSlasher is SlasherUtil {
-      "Signers do not match"
+      startBlocks.length == endBlocks.length,
+      "startBlocks and endBlocks must have the same length"
     );
-    uint256 sz = getEpochSize();
-    uint256 startEpoch = epochNumberOfBlock(startBlock, sz);
-    for (uint256 n = startBlock; n <= endBlock; n = n.add(1)) {
-      uint256 signerIndex = epochNumberOfBlock(n, sz) == startEpoch
-        ? startSignerIndex
-        : endSignerIndex;
-      // We want to check signers for block n,
-      // so we get the parent seal bitmap for the next block
-      if (uint256(getParentSealBitmap(n.add(1))) & (1 << signerIndex) != 0) return false;
+    require(signerIndices.length > 0, "requires at least one signerIndex");
+
+    uint256 epochSize = getEpochSize();
+    uint256 signerIndicesIndex = 0;
+    for (uint256 i = 0; i < startBlocks.length; i = i.add(1)) {
+      if (i > 0) {
+        require(
+          startBlocks[i.sub(1)] < startBlocks[i],
+          "each interval must start after the previous interval's start"
+        );
+        require(
+          startBlocks[i] <= endBlocks[i.sub(1)].add(1),
+          "each interval must start at least one block after previous interval's end"
+        );
+        require(
+          endBlocks[i.sub(1)] < endBlocks[i],
+          "each interval must end after the previous interval"
+        );
+        // The signer index of a particular validator may change from epoch to epoch.
+        // Because the intervals for which bitmaps are calculated in this contract do not span
+        // epochs, and because intervals processed by this function are guaranteed to be
+        // overlapping or contiguous, whenever we cross epoch boundaries we are guaranteed to
+        // process an interval that starts with the first block of that epoch.
+        if (startBlocks[i].sub(1).mod(epochSize) == 0) {
+          require(
+            getValidatorAccountFromSignerIndex(
+              signerIndices[signerIndicesIndex],
+              startBlocks[i].sub(1)
+            ) ==
+              getValidatorAccountFromSignerIndex(
+                signerIndices[signerIndicesIndex.add(1)],
+                startBlocks[i]
+              ),
+            "indices do not point to the same validator"
+          );
+          signerIndicesIndex = signerIndicesIndex.add(1);
+        }
+      }
+      if (!wasDownForInterval(startBlocks[i], endBlocks[i], signerIndices[signerIndicesIndex])) {
+        return false;
+      }
     }
+
     return true;
   }
 
   /**
-   * @notice Returns the end block for the interval.
+   * @notice Add to the validator a new lastSlashedBlock if the validator was not already slashed
+   * in the same epoch of the startBlock and if the blocks to slash are newer than the last one
    */
-  function getEndBlock(uint256 startBlock) internal view returns (uint256) {
-    return startBlock.add(slashableDowntime).sub(1);
-  }
+  function addValidatorSlash(address validator, uint256 startBlock, uint256 endBlock) internal {
```

Comment:
Consider folding this logic into `slash`, and simplifying it to the following:

```

require(startBlocks[0] >= lastSlashedBlock[validator], "cannot slash validator for downtime for which they may already have been slashed");
require(wasDownForIntervals(startBlocks, endBlocks, signerIndices), "not down");
lastSlashedBlock[validator] = endBlocks[-1];
performSlashing();
```

IMO there's no benefit to requiring validators cannot get slashed twice for downtime within the same epoch.

Reply:
There was a flag in the beginning that allows us to slash more than once per epoch and was required to be changed:
https://github.com/celo-org/celo-monorepo/pull/3806#discussion_r435507918
and this one:
https://github.com/celo-org/celo-monorepo/pull/3806#discussion_r430740798

I think that it was required to be removed mostly because to make the code clearer.

But the possibility of doing that, it was also part of a discussion in slack.
And I remember that it was a split decision, haha.
Basically:
(the argument of the once per epoch): "If you messed up with something that means your node is down all day, you only ought be able to be slashed once for that. After all, the idea is mainly to force your validator to get ejected from the group, which happens on the first slashing."
(one of the arguments of the n per epoch): "if one validator was down the first 8 hours, there is no need to even try to get the node running again. At least with the possibility of having another slash, could make that happen"

Honestly I prefer to allow more than once, but was something that it was required to change. Maybe now that everything it's clearer, we could allow it

Reply:
Changed to allow it